### PR TITLE
Add Silex provider

### PR DESCRIPTION
**Pull request by thewilkybarkid**

### CODE REVIEW

File: src/Silex/LoggerProvider.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                    $e instanceof HttpExceptionInterface && $e->getStatusCode() < 500
+                    ||
+                    $e instanceof ApiProblemException && $e->getStatusCode() < 500
+                ) {
+                    return LogLevel::INFO;
+                }
+
+                return LogLevel::CRITICAL;
+            });
+        };
```
Comment (giorgiosironi) [inline src/Silex/LoggerProvider.php]: good direction, can this include the `error()` Silex setup?

File: composer.json [hunk 2]
```diff
@@ -19,10 +19,14 @@
         "psr/log": "^1.0"
     },
     "require-dev": {
-        "phpunit/phpunit": "^5.7 || ^6.0"
+        "elife/api-problem": "^1.0",
```
Comment (giorgiosironi) [inline composer.json]: a bit strange that a logging library also brings in an API library (you could log without having an API); as long as it's not circular, not an actual problem

File: src/Silex/LoggerProvider.php (hunk 1 above)
Comment (thewilkybarkid) [reply src/Silex/LoggerProvider.php]: The listener happens before Silex error handlers.

File: composer.json (hunk 2 above)
Comment (thewilkybarkid) [reply composer.json]: I would make `ApiProblemException` implement `HttpExceptionInterface`, but it comes from Symfony's HttpKernel. (Unlikely we would ever not use that, but it's an extract dependency...)

**Comment by thewilkybarkid** [comment]
Looking at testing. Currently forced to write files, wondering if we can capture the logs (relates to #4).

**Comment by thewilkybarkid** [comment]
Trying vfsStream.

File: test/Silex/LoggerProviderTest.php [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    public function createApplication() : Application
+    {
+        $this->logs = vfsStream::setup('logs');
+
+        $app = new Application();
+        $app['logger.channel'] = 'app';
+        $app['logger.level'] = LogLevel::INFO;
+        $app['logger.path'] = vfsStream::url($this->logs->getName());
```
Comment (giorgiosironi) [inline test/Silex/LoggerProviderTest.php]: it's not the end of the world if we have to set up some real test folders, but it's elegant to do everything in memory

File: src/Silex/LoggerProvider.php (hunk 1 above)
Comment (giorgiosironi) [reply src/Silex/LoggerProvider.php]: I take it this means we don't need to setup `error()`, which is what I was after 👍

Comment (thewilkybarkid) [reply src/Silex/LoggerProvider.php]: Yep, https://github.com/silexphp/Silex/blob/e970d12a5b5a203fbd859525ce406b22e73f9f61/src/Silex/EventListener/LogListener.php#L127-L131

File: src/Silex/LoggerProvider.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+        $app['logger'] = function () use ($app) {
+            return $app['logger.factory']->logger();
+        };
+
+        $app['logger.listener'] = function () use ($app) {
+            return new LogListener($app['logger'], function (Throwable $e) {
+                if (
+                    $e instanceof HttpExceptionInterface && $e->getStatusCode() < 500
```
Comment (giorgiosironi) [inline src/Silex/LoggerProvider.php]: depends on the precedence of `&&` over `||`, a couple of parentheses would make it more unambiguous

File: src/Silex/LoggerProvider.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                    ||
+                    $e instanceof ApiProblemException && $e->getApiProblem()->getStatus() < 500
+                ) {
+                    return LogLevel::INFO;
+                }
+
+                return LogLevel::CRITICAL;
```
Comment (giorgiosironi) [inline src/Silex/LoggerProvider.php]: `ERROR` could probably be enough? At least that's the de-facto standard, but we don't have an ADR for that

**Comment by thewilkybarkid** [comment]
:shipit:?

**Pull request merged**